### ISSUE TITLE
I can't understand something in function ContentLoss:updateGradInput(input, gradOutput)

### ISSUE THREAD

**Issue by mathfinder** [jcjohnson/neural-style#387]
In function ContentLoss:updateGradInput(input, gradOutput),what's gradOutput's mean and why the gradInput should and gradOutput?

**Comment by htoyryla** [comment]
> and why the gradInput should and gradOutput?

If you mean "and why the gradInput should add gradOutput", I have understood that it is in norder to accumulate the gradients from all loss modules so that the optimizer can change the image accordingly.

**Comment by SSeanHsu** [comment]
@htoyryla I also don't really get the aim of add gradOutput. Is that a usual way to do so in the function of updateGradInput? Can you add more details?

**Comment by htoyryla** [comment]
I am not at all an expert in backpropagation but still learning. Anyhow, this is how it appears to me. What neural-style does is not like the usual backpropagation where one evaluates loss at the output of the model and then backpropagates it down the network. 

Here, we evaluate losses at several places inside the network, and if a loss module would not pass downwards the effect of the loss modules above it, the effect of those loss modules would be ignored. Only the lowest loss module would have any effect. Note also that we are not training the model here, what matters is the gradients that reach the bottom and are used to modify the input.

**Comment by SSeanHsu** [comment]
@htoyryla I have two questions. And I am going to attach Jcjohnson's original codes.
`function ContentLoss:updateGradInput(input, gradOutput)
  if self.mode == 'loss' then
    if input:nElement() == self.target:nElement() then
      self.gradInput = self.crit:backward(input, self.target)
    end
    if self.normalize then
      self.gradInput:div(torch.norm(self.gradInput, 1) + 1e-8)
    end
    self.gradInput:mul(self.strength)
    self.gradInput:add(gradOutput)
  else
    self.gradInput:resizeAs(gradOutput):copy(gradOutput)
  end
  return self.gradInput
end`
My questions are as follows,
1. If in one situation which the self.mode == 'loss', neither input:nElement() == self.target:nElement() nor self.normalize == ture, the code would only execute `self.gradInput:mul(self.strength)
    self.gradInput:add(gradOutput)`. So in this situation, I haven't compute self.gradInput yet. How would I know what will multiply self.strength and add gradOutput? 
2. I am little confused as the original codes set `-content_layers` `relu4_2`, `self.gradInput = self.crit:backward(input, self.target)` would execute twice. But I don't really know why we need to execute this twice?

**Comment by htoyryla** [comment]
1.

In a ContentLoss module, before we enter loss mode, the target should have been set in capture mode from the input (see ContentLoss:updateOutput) so the both target and input should have the same number of elements. If they don't then it makes no sense to compare them either with the criterion.

2.
I don't this at all. Where is this and what has the -content_layer parameter have to do here?

**Comment by SSeanHsu** [comment]
So do you mean `input:nElement() == self.target:nElement()` is always ture? So that the code `if input:nElement() == self.target:nElement() then
      self.gradInput = self.crit:backward(input, self.target)
    end` is actually insuring that we have correct size of input and target?

**Comment by htoyryla** [comment]
If the target has been captured correctly then 
```
input:nElement() == self.target:nElement()
```
should always be true. If it isn't, the module cannot add anything to the gradient we get from output, and then we just set gradInput = gradOutput.

What a loss module does in neural-style is to capture a target, and during iterations compute the difference between the target and the  current value (= input of the module for content, gram matrix output for style) and add that to the gradient from above. In this manner we finally get at the bottom of the model a total gradient pointing out how the image should be modified.

**Comment by SSeanHsu** [comment]
Got it. Thank you so much for your patient explaination :D
